src/Field/BaseFieldDescription.php

```diff
@@ -0,0 +1,326 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\DatagridBundle\Field;
+
+use Doctrine\Inflector\InflectorFactory;
+use Sonata\DatagridBundle\Exception\NoValueException;
+
+/**
+ * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ */
+abstract class BaseFieldDescription implements FieldDescriptionInterface
+{
+    /**
+     * @var string|null the field name
+     */
+    protected $name;
+
+    /**
+     * @var string|null the field name (of the form)
+     */
+    protected $fieldName;
+
+    /**
+     * @var string|int|null the type
```

Comment:
I just basically moved the file from the AdminBundle to the DatagridBundle
https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/Admin/BaseFieldDescription.php#L70

I don't know why int. But it was allowed before.

Reply:
I add explanation. For example, ClassMetadata::ONE_TO_MANY is allowed.

Reply:
This looks like an architecture problem when should fix when moving this to the right place.

The `type` property name is to generic to guess the right value. Someone could misuse this.

Reply:
> This looks like an architecture problem

Maybe, this seems to works well like this currently.

After some investigation, 
the type is a string like `orm_many_to_one`

The only reason to allow `int` was because of the implementation
```
public function setAssociationMapping($associationMapping)
    {
        if (!\is_array($associationMapping)) {
            throw new \RuntimeException('The association mapping must be an array');
        }

        $this->associationMapping = $associationMapping;

        $this->type = $this->type ?: $associationMapping['type'];
        $this->mappingType = $this->mappingType ?: $associationMapping['type'];
        $this->fieldName = $associationMapping['fieldName'];
    }
```

Maybe this implementation is wrong and we should remove the line
```
$this->type = $this->type ?: $associationMapping['type'];
```

Maybe we should also makes field private in order to force the usage of getter/setter.
I'll do it tonight.

> We should fix when moving this to the right place.

I disagree. 

Keep in mind there is mutiple of issue we could fix for datagrid, field descriptions and more.
I can't fix them all. 

Fixing some of them is still better than fixing nothing.
I would prefer to focus about the duplicated code between the two bundles.

Reply:
After some more investigation, there some times where the type is set as an integer (for orm mapping), when I remove this behaviour, the type is null.

The type is use in DoctrineORMAdmin
```
$fieldDescription->setTemplate($this->getTemplate($fieldDescription->getType()));
```
If no template is found (which occur for 1, 2, 3 or null), it then use the Mapping type.

A possible solution would be to restrict type to `?string` and mapping type to `?int`.
And we'll have to stop throwing error when type is not found.

WDYT @core23 ?